cpp/src/io/utilities/datasource.cpp

```diff
@@ -59,19 +69,31 @@ class file_source : public datasource {
 
   std::unique_ptr<buffer> host_read(size_t offset, size_t size) override
   {
-    // Clamp length to available data
-    auto const read_size = std::min(size, this->size() - offset);
-    std::vector<uint8_t> v(read_size);
-    CUDF_EXPECTS(_kvikio_file.pread(v.data(), read_size, offset).get() == read_size, "read failed");
-    return buffer::create(std::move(v));
+    auto clamped_read = clamped_read_to_vector(offset, size);
+    clamped_read.second.get();
+    return buffer::create(std::move(clamped_read.first));
+  }
+
+  std::future<std::unique_ptr<datasource::buffer>> host_read_async(size_t offset,
+                                                                   size_t size) override
+  {
+    auto clamped_read = clamped_read_to_vector(offset, size);
+    return std::async(std::launch::deferred, [cr = std::move(clamped_read)]() mutable {
+      cr.second.get();
+      return buffer::create(std::move(cr.first));
+    });
   }
 
   size_t host_read(size_t offset, size_t size, uint8_t* dst) override
+  {
+    return host_read_async(offset, size, dst).get();
+  }
+
+  std::future<size_t> host_read_async(size_t offset, size_t size, uint8_t* dst) override
   {
     // Clamp length to available data
     auto const read_size = std::min(size, this->size() - offset);
-    CUDF_EXPECTS(_kvikio_file.pread(dst, read_size, offset).get() == read_size, "read failed");
-    return read_size;
+    return _kvikio_file.pread(dst, read_size, offset);
```

Comment:
Naive question: could we use the Kvikio's `read` for `host_read` and `pread` for `host_read_async`?

Reply:
No, it's a good question :)
We use pread because it breaks large reads into slices and performs them in parallel. so we potentially get faster even when we serially call preads and immediately get the result.